File: python/day4.py
"""
Atanas Delevski
12/20/2022
Advent of Code 2022 - Day 4 (Part I & II?)
"""


def part_one(lines):
    ans = 0
    for line in lines:
        f, s = line.split(',')
        f_f, f_s = [int(x) for x in f.split('-')]
        s_f, s_s = [int(x) for x in s.split('-')]
        if (s_f <= f_f <= s_s) and (s_f <= f_s <= s_s):
            ans += 1
        elif (f_f <= s_f <= f_s) and (f_f <= s_s <= f_s):
            ans += 1
    return ans


def part_two(lines):
    ans = 0
    for line in lines:
        f, s = line.split(',')
        f_f, f_s = [int(x) for x in f.split('-')]
        s_f, s_s = [int(x) for x in s.split('-')]
        if (s_s < f_f) or (s_f > f_s):
            continue
        else:
            ans += 1
    return ans


def main():
    # Init
    file_path = 'data/day4.txt'
    file = open(file_path, "r")
    lines = file.readlines()

    # Part 1:
    part_one_answer = part_one(lines)

    # Part 2:
    part_two_answer = part_two(lines)

    # Printing
    print(f"Part 1: {part_one_answer}")
    print(f"Part 2: {part_two_answer}")


if __name__ == "__main__":
    main()

File: src/main/scala/day4.scala
import scala.io.Source



def proc(x: String): Int = {
    val a  = x.split(",|-").map(_.toInt)
    if ((a(0) <= a(2) && a(3) <= a(1)) ||
    (a(2) <= a(0) && a(1) <= a(3))) 1 else 0
}


def proc2(x: String): Int = {
    val a = x.split(",|-").map(_.toInt)
    if ((a(1) < a(2)) ||
        (a(0) > a(3))) 0 else 1
}


@main def day4() = {
    val lines: List[String] = Source.fromFile("data/day4.txt").getLines.toList
    println(s"Part 1: ${lines.map(proc).sum}")
    println(s"Part 2: ${lines.map(proc2).sum}")
}

